Update limb_darkening_fit.py to correct matplotlib/TeX errors

AarynnCarter opened the pull request:
I created this branch to directly address issue #42 where matplotlib/TeX causes crashes at a variety of different stages in the pipeline depending on the user. 

After some testing, I'm almost certain that the following lines on one of the Stage 5 files has been editing the underlying matplotlib rcParams whenever Eureka is imported, and is directly causing these issues. I've attached some images below of what a plot looks like with and without these lines. Personally, I actually prefer the default matplotlib font, although we might need to make some changes to the font sizes directly in our plotting routines. 

```
rc('font', **{'family': 'sans-serif', 'sans-serif': ['Helvetica'], 'size': 16})
rc('text', usetex=True)
```

@taylorbell57 Could you install this branch and remove the additional lines you added to the jwst package file to see if the previous error disappears for you? The same issue cropped up again for me this morning (spurring me to try and fix things once and for all) and removing these two lines fixed things. 

@kevin218 @lkreidberg Changing the overall plotting style for Eureka seems pretty significant, but I think it's worth it for broader usability. Given you're leading things I'll let you decide whether to merge this or not!

![ERS_DATASIMv2_NIRSpec_PRISM_TINY_x1dints](https://user-images.githubusercontent.com/23636747/138347120-1a9b1a31-94a9-4fe3-af44-4c05e6fb2673.png)
![ERS_DATASIMv2_NIRSpec_PRISM_TINY_x1dints](https://user-images.githubusercontent.com/23636747/138347197-376a9563-1746-403b-abfd-c374fce68911.png)

taylorbell57 commented:
I agree that having users directly edit the jwst file doesn't make sense, but contributing that change to the `jwst` package and requiring `eureka` to use that newer version might be a good option. But then we might run into issues with which `jwst` version is compatible with the current simulated observations (thankfully that'll be resolved once we actually get some real flight data). I am open to also not requiring tex fonts without an opt-out feature (not everyone will want to install tex with things like overleaf now), but I think it's best to resolve the conflict now rather than side-step it and let others figure out how to solve it down the road if they want to use a nicer font. Either way, setting this in a Stage 5 file that then gets applied to all scripts is definitely not an intuitive or sensible decision, so I suggest we move the two lines you cut to another file somewhere more obvious and add an option to skip them somehow. I'm not sure a .ecf option is the cleanest way though

AarynnCarter commented:
Yes, I suppose there are a few overlapping issues here. I totally agree that we should provide the solution to the jwst pipeline team and update Eureka to use that new version or higher, in fact, I'll do this now.  

Maybe this would be something to discuss on Monday, but I think it's important to decide between 1) opting-in to TeX fonts, or 2) opting-out of TeX fonts. I think 1) is the better choice, as making installation and code execution as hassle-free as possible seems more valuable to me. Like you say, we definitely shouldn't have these lines a Stage 5 file, my immediate thought is that we move them to the lib/plots.py file under a toggle-able parameter. For example... 

```
from eureka.lib import plots
plots.use_tex_fonts = True
plots.family = 'sans-serif'
plots.font = 'Helvetica'
```

taylorbell57 commented:
I 100% agree to opt-in to tex fonts - opt-out is definitely going to cause more people headaches than slightly less nice plots will

AarynnCarter commented:
Okay, it's a little different to what I initially put, but to opt in for TeX fonts you can now add to whatever code you're running:

```
from eureka.lib import plots
plots.use_tex_fonts()
```

At the moment you can also pass a `family`, `font`, and `fontsize` argument to this for more manual customisation. Should be relatively easy to expand to other rc parameters if and when we need to.

kevin218 commented:
It doesn't seem like there's a way to set Tex fonts to False. A boolean argument would work to toggle it on or off.

I like the idea of making this a general function for defining the rc parameters. This will make it possible to generate consistent, paper-ready plots. If that's the plan, let's rename the routine.

AarynnCarter commented:
Good catch Kevin, I didn't think about that. The new function takes a boolean argument to switch to a TeX font, or switch back to the matplotlib default. 

Are you thinking something like: `plots.make_paper_ready(True)`?

kevin218 commented:
I was thinking `plots.set_rc(bunch of stuff)`

taylorbell57 commented:
A **kwargs would be perfect for that

taylorbell57 commented:
Hi all, what do you think about this instead?

**Current Code:**
```
def use_tex_fonts(toggle, family='sans-serif', font='Helvetica', fontsize=16):
    if toggle == True:
        rc('font', **{'family': family, family: [font], 'size': fontsize})
        rc('text', usetex=True)
    elif toggle == False:
        rcdefaults()
    else: 
        raise ValueError('TeX font toggle must be a boolean.')
```

**My Suggestion:**
```
def set_rc(user_custom=False, usetex=False, **kwargs):
    if not (isinstance(user_custom, bool) and isinstance(usetex, bool)):
        raise ValueError('user_custom and usetex arguments must be boolean.')

    if user_custom:
        # Let users do whatever they'd like
        rcdefaults()
        rc('font', **kwargs)
    else:
        # Fall back to some nice Eureka! recommended font settings
        family='sans-serif'
        font='Helvetica'
        fontsize=16
        rc('font', **{'family': family, family: [font], 'size': fontsize})

    rc('text', usetex=usetex)
```

If needed in the future, this could be further generalized upon to allow Eureka! customization to line properties and whatnot with separate font_kwargs, line_kwargs, xtick_kwargs, etc.

AarynnCarter commented:
Hey @taylorbell57, that looks much better! **kwargs are definitely the way forward here. It also looks pretty straightforward to expand things if we want multiple default plotting styles. 

If @kevin218 is happy, I can test/add this in today and merge with main.

kevin218 commented:
Go for it!

AarynnCarter commented:
Okay, so I ended up tweaking things a little more to cover a few edge cases:

1) If we were to toggle on the Eureka settings, it didn't seem intuitive to me to call `set_rc(user_custom=True) ` to reapply the default matplotlib rcParams. Instead, I've switched things to a more specific `style` parameter.

2) If we were to set some `user_custom` style, but then decide later on in some code to switch the `usetex` setting, it would wipe all of the applied settings, unless they were provided once over. To avoid this, I've added a `preserve` style to simply not adjust the rcParams at all. 

3) Again for the `user_custom` case, if the function was called multiple times it might wipe some previously applied settings. So I've moved the `rcdefaults()` call to a `from_scratch` boolean so that users have the option to reset the rcParams entirely, or apply further changes. 

If we're happy with this I just need someone to approve and I can merge. 

```
# Function to adjust matplotlib rcParams for plotting procedures.
def set_rc(style='preserve', usetex=False, from_scratch=False, **kwargs):
    if not (isinstance(usetex, bool) and isinstance(from_scratch, bool)):
        raise ValueError('"usetex" and "from_scratch" arguments must be boolean.')

    if from_scratch:
        # Reset all rcParams prior to making changes. 
        rcdefaults()

    # Apply the desired style...
    if style == 'custom':
        # Custom user font, kwargs must be from the 'font' rcParams group at the moment. 
        # https://matplotlib.org/stable/api/matplotlib_configuration_api.html#matplotlib.rcParams
        rc('font', **kwargs)
    elif style == 'eureka':
        # Apply default Eureka! font settings.
        family='sans-serif'
        font='Helvetica'
        fontsize=16
        rc('font', **{'family': family, family: [font], 'size': fontsize})
    elif style == 'default':
        # Use default matplotlib settings
        rcdefaults()
    elif style == 'preserve':
        pass
    else:
        raise ValueError('Input style must be one of: "custom", "eureka", "preserve", or "default".')

    rc('text', usetex=usetex) # TeX fonts may not work on all machines. 
```
